lib/smack/NormalizeLoops.cpp

```diff
@@ -0,0 +1,116 @@
+//
+// This file is distributed under the MIT License. See LICENSE for details.
+//
+
+//
+// This pass normalizes loop structures to allow easier generation of Boogie
+// code when a loop edge comes from a branch. Specifically, a forwarding block
+// is added between a branching block and the loop header.
+//
+
+#include "smack/NormalizeLoops.h"
+#include "smack/Naming.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/ValueSymbolTable.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+#include <set>
+
+namespace smack {
+
+using namespace llvm;
+
+// Register LoopInfo
+void NormalizeLoops::getAnalysisUsage(AnalysisUsage& AU) const {
+  AU.addRequired<LoopInfoWrapperPass>();
+}
+
+BasicBlock* mkForwardingBlock(BasicBlock* target, Function* func) {
+  auto& context = target->getContext();
+  BasicBlock* result = BasicBlock::Create(context, "forwarder", func);
+  BranchInst* branch = BranchInst::Create(target);
+  result->getInstList().push_back(branch);
+  return result;
+}
+
+void processPhis(std::vector<PHINode*>& phis, std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  for (PHINode* phi : phis) {
+    unsigned numIncoming = phi->getNumIncomingValues();
+    for (unsigned i = 0; i < numIncoming; ++i) {
+      BasicBlock* incomingBlock = phi->getIncomingBlock(i);
+      if (blockMap.count(incomingBlock) == 1) {
+        phi->setIncomingBlock(i, blockMap[incomingBlock]);
+      }
+    }
+  }
+}
+
+void processLoop(Loop *loop, Function* func) {
+  std::vector<PHINode*> phis;
+  std::set<BasicBlock*> phiIncBlocks;
+  BasicBlock* headerBlock = loop->getHeader();
+
+  // Gather Phis in the header
+  for (Instruction& I : *headerBlock) {
+    PHINode* phi = dyn_cast<PHINode>(&I);
+    if (phi) {
+      phis.push_back(phi);
+      phiIncBlocks.insert(phi->block_begin(), phi->block_end());
+    }
+  }
+
+  std::map<BasicBlock*, BasicBlock*> blockMap;
+
+  // Add forwarding blocks
+  for (BasicBlock* BB : phiIncBlocks) {
```

Comment:
Pull this into a separate function since this one is pretty big.

Reply:
Done.